### PR TITLE
ftx fetchAllBalances

### PR DESCRIPTION
**Pull request by samgermain**
Returns a response for `ftx.privateGetWalletAllBalances` that more closely resembles our unified response for fetchBalance. Currently this is a draft because the bottom two values for `free` and `total` show undefined

---------------------

```
2021-11-27T03:17:14.364Z
Node.js: v14.17.0
CCXT v1.62.38
ftx.fetchAllBalances ()
{  info: { success:    true,
            result: { main: [ {                   coin: "GT",
                                                 total: "0.0",
                                                  free: "0.0",
                                availableWithoutBorrow: "0.0",
                                              usdValue: "0.0",
                                            spotBorrow: "0.0"  },
                              {                   coin: "LTC",
                                                 total: "0.0",
                                                  free: "0.0738798",
                                availableWithoutBorrow: "0.0",
                                              usdValue: "0.0",
                                            spotBorrow: "0.0"        },
                              {                   coin: "CAD",
                                                 total: "0.0",
                                                  free: "18.89854238",
                                availableWithoutBorrow: "0.0",
                                              usdValue: "0.0",
                                            spotBorrow: "0.0"          },
                              {                   coin: "FTT",
                                                 total: "0.0",
                                                  free: "0.0",
                                availableWithoutBorrow: "0.0",
                                              usdValue: "0.0",
                                            spotBorrow: "0.0"  },
                              {                   coin: "ETH",
                                                 total: "0.0",
                                                  free: "0.00356206",
                                availableWithoutBorrow: "0.0",
                                              usdValue: "0.0",
                                            spotBorrow: "0.0"         },
                              {                   coin: "USD",
                                                 total: "1.68859901",
                                                  free: "14.95396208",
                                availableWithoutBorrow: "0.81459901",
                                              usdValue: "1.688599018101419",
                                            spotBorrow: "0.0"                },
                              {                   coin: "BTC",
                                                 total: "0.0",
                                                  free: "0.00027016",
                                availableWithoutBorrow: "0.0",
                                              usdValue: "0.0",
                                            spotBorrow: "0.0"         },
                              {                   coin: "USDT",
                                                 total: "48.32564944",
                                                  free: "20.15920217",
                                availableWithoutBorrow: "20.15920217",
                                              usdValue: "48.35481155315457",
                                            spotBorrow: "0.0"                }  ] } },
   main: {    GT: { free: "0.0", used: undefined, total: "0.0" },
             LTC: { free: "0.0", used: undefined, total: "0.0" },
             CAD: { free: "0.0", used: undefined, total: "0.0" },
             FTT: { free: "0.0", used: undefined, total: "0.0" },
             ETH: { free: "0.0", used: undefined, total: "0.0" },
             USD: {  free: "0.81459901",
                     used:  undefined,
                    total: "1.68859901"  },
             BTC: { free: "0.0", used: undefined, total: "0.0" },
            USDT: {  free: "20.15920217",
                     used:  undefined,
                    total: "48.32564944"  },
            free:    undefined,
            used:    undefined,
           total:    undefined                                    },
   free: { main: undefined },
   used: { main: undefined },
  total: { main: undefined }                                                           }
575 ms
```

### CODE REVIEW

File: js/ftx.js [hunk 1]
```diff
@@ -1121,6 +1121,65 @@ module.exports = class ftx extends Exchange {
...
+            const accountName = accounts[i];
+            const balances = data[accountName];
+            returnObject[accountName] = {};
+            for (let j = 0; j < balances.length; j++) {
+                const balance = balances[j];
+                const code = this.safeCurrencyCode (this.safeString (balance, 'coin'));
+                const account = this.account ();
+                account['free'] = this.safeString2 (balance, 'availableWithoutBorrow', 'free');
+                account['total'] = this.safeString (balance, 'total');
+                returnObject[accountName][code] = account;
```
Comment (kroitor) [inline js/ftx.js:1177]: Logically, we have to do the following here:

```JavaScript
        balance[code] = account;
    }
    result[accountId] = parseBalance (balance);
}
return result;
```

Just like we do here:

- https://github.com/ccxt/ccxt/blob/master/js/huobi.js#L1977
- https://github.com/ccxt/ccxt/blob/master/js/okex3.js#L1640

Comment (samgermain) [reply js/ftx.js:1177]: @kroitor balance is an object from the api response, so it doesn't make sense to parse that object, I'm not exactly sure what you're trying to describe?

Comment (samgermain) [reply js/ftx.js:1177]: @kroitor I'm just waiting on an answer here

Comment (kroitor) [reply js/ftx.js:1177]: @samgermain i'll explain it in more details asap

Comment (samgermain) [reply js/ftx.js:1177]: @kroitor Do you have any more details on this?

Comment (kroitor) [reply js/ftx.js:1177]: @samgermain in your initial output the values of `free`, `used`, `total` are all undefined. I think `parseBalance` has to be applied to each of the accounts and the entire fetchAllBalances method should return parsed numbers.

**Pull request closed without merging**